mixer/adapter/redisquota/redisquota.go

```diff
@@ -0,0 +1,509 @@
+	ac := b.adapterConfig
+
+	limits := make(map[string]*config.Params_Quota, len(ac.Quotas))
+	for idx := range ac.Quotas {
+		l := ac.Quotas[idx]
+		limits[l.Name] = &l
+	}
+
+	for k := range b.quotaTypes {
+		if _, ok := limits[k]; !ok {
+			return nil, env.Logger().Errorf("did not find limit defined for quota %s", k)
+		}
+	}
+
+	option := redis.Options{
+		Addr: b.adapterConfig.RedisServerUrl,
+	}
+
+	if b.adapterConfig.ConnectionPoolSize > 0 {
+		option.PoolSize = int(b.adapterConfig.ConnectionPoolSize)
+	}
+
+	// initialize client
+	client := redis.NewClient(&option)
+
+	// test connection
+	_, err := client.Ping().Result()
+	if err != nil {
+		return nil, env.Logger().Errorf("could not create a connection pool with redis: %v", err)
+	}
+
+	// load script
+	scripts := map[string]*redis.Script{}
+	for algorithm, script := range rateLimitingLUAScripts {
+		scripts[algorithm] = redis.NewScript(script)
+		_, err := scripts[algorithm].Load(client).Result()
+		if err != nil {
+			return nil, env.Logger().Errorf("unable to load the LUA script: %v", err)
+		}
+	}
+
+	h := &handler{
+		client:  client,
+		limits:  limits,
+		scripts: scripts,
+		logger:  env.Logger(),
+		getTime: time.Now,
+	}
+
+	return h, nil
+}
+
+////////////////// Runtime Methods //////////////////////////
+
+// matchDimensions matches configured dimensions with dimensions of the instance.
+func matchDimensions(cfg map[string]string, inst map[string]interface{}) bool {
+	for k, val := range cfg {
+		rval := inst[k]
+		if rval == val { // this dimension matches, on to next comparison.
+			continue
+		}
+
+		// if rval has a string representation then compare it with val
+		// For example net.ip has a useful string representation.
+		switch v := rval.(type) {
+		case fmt.Stringer:
+			if v.String() == val {
+				continue
+			}
+		}
+		// rval does not match val.
+		return false
+	}
+	return true
+}
+
+func getAllocatedTokenFromResult(result interface{}) (int64, time.Duration, error) {
+	if res, ok := result.([]interface{}); ok {
+		if reflect.ValueOf(result).Len() != 2 {
+			return 0, time.Duration(0), fmt.Errorf("invalid response from the redis server: %v", result)
+		}
+
+		// read token
+		tokenValue, tokenOk := res[0].(int64)
+		if !tokenOk {
+			return 0, time.Duration(0), fmt.Errorf("invalid response from the redis server: %v", result)
+		}
+
+		// read expiration
+		expValue, expOk := res[1].(int64)
+		if !expOk {
+			return 0, time.Duration(0), fmt.Errorf("invalid response from the redis server: %v", result)
+		}
+
+		return int64(tokenValue), time.Duration(expValue) * time.Nanosecond, nil
+	}
+	return 0, time.Duration(0), fmt.Errorf("invalid response from the redis server: %v", result)
+}
+
+// getOverrideHash returns hash kye of the given dimension in sorted by key
+func getDimensionHash(dimensions map[string]string) string {
+	var keys []string
+	for k := range dimensions {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	h := fnv.New32a()
```

Comment:
This ends up generating a fair bit of garbage on every use. Any way this can be done in reusable pooled buffers?

Reply:
Created a map of memory address to hash on initialization and HandleQuota lookup the map with the address of matching Dimensions